api/src/main/java/com/datatorrent/api/Operator.java

```diff
@@ -270,4 +271,29 @@ public String toString()
 
   }
 
+  /**
+   * Operators that need to be notified about checkpoint events should implement this interface.
+   *
+   * The notification callbacks in this interface are called outside window boundaries so the operators should not
+   * attempt to send any tuples in these callbacks.
+   *
+   */
+  interface CheckpointNotificationListener extends CheckpointListener
+  {
+    /**
+     * Notify the operator that a checkpoint is about to be performed.
+     *
+     * Operators may need to perform certain tasks just before a checkpoint such as calling flush on a stream. Having
+     * this notification helps operators perform such operations optimally by doing them once before checkpoint as
+     * opposed to doing it repeatedly at the end of every window.
+     *
+     * The method will be called before the checkpoint is performed. It will be called after
+     * {@link Operator#endWindow()} call of the window preceding the checkpoint and before the checkpoint is
+     * actually performed.
+     *
+     * @param windowId The window id of the window preceding the checkpoint
+     */
+    void prepareCheckpoint(long windowId);
```

Comment:
@PramodSSImmaneni : The name suggests that every Stateless Operator needs to implement this which is not true.